text/0000-cfg-target-abi.md

```diff
@@ -0,0 +1,112 @@
+- Feature Name: `cfg-target-abi`
+- Start Date: 2020-09-27
+- RFC PR: [rust-lang/rfcs#2992](https://github.com/rust-lang/rfcs/pull/2992)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This proposes a new `cfg`: `target_abi`, which specifies certain aspects of the
+target's [Application Binary Interface (ABI)][abi]. This also adds a
+`CARGO_CFG_TARGET_ABI` environment variable for parity with other
+`CARGO_CFG_TARGET_*` variables.
+
+# Motivation
+[motivation]: #motivation
+
+Certain targets are only differentiated by their ABI. For example: the `ios` OS
+in combination with the `macabi` ABI denotes targeting Mac Catalyst (iOS on
+macOS). The non-`macabi` `x86_64-apple-ios` target is not for Mac Catalyst and
+instead is for the iOS simulator, which is a very different environment.
+
+It is not currently possible to `#[cfg]` against a certain target ABI without
+a `build.rs` script to emit a custom `cfg` based on the `TARGET` environment
+variable. This is not ideal because:
+
+- Adding a build script increases compile time and makes a crate incompatible
+  with certain build systems.
+
+- Checking `TARGET` is error prone, mainly because the ABI often follows
+  `target_env` without separation.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This would act like [existing `target_*` configurations][cfg-options].
+
+For example: if one had a module with bindings to
+[Apple's AppKit](https://developer.apple.com/documentation/appkit), this feature
+could be used to ensure the module is available when targeting regular macOS and
+Mac Catalyst.
+
+```rust
+#[cfg(any(
+    target_os = "macos",
+    all(
+        target_os = "ios",
+        target_abi = "macabi",
+    ),
+))]
+pub mod app_kit;
+```
+
+This configuration option would also be usable as
+`#[cfg_attr(target_abi = "...", attr)]`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`target_abi` is a key-value option set once with the target's ABI. The value is
```

Comment:
Does rustc already have the concept of target ABI internally? If not it seems really important to spell out how exactly this is defined. The concept of ABIs is really poorly understood by developers in my experience. Obviously there's a 1:1 mapping of target : ABI, but the inverse mapping is a little fuzzy to me and I'm not sure how useful it is in practice.

For example, I'm not really aware of an existing convention for talking about the "x86-64 System V ABI" despite it being the ABI used in both x86-64 Linux and macOS.

Reply:
Specifically what I'm interested in is the component that comes after `target_env` in the triple string. What it indicates varies greatly between different targets. On some targets, it indicates the availability of hardware floats (e.g. `eabihf` for **e**mbedded **abi**, **h**ardware **f**loat). On others, it indicates the running environment (e.g. `x86_64-apple-ios-macabi` for iOS apps that target macOS).

Reply:
To answer my own question, yes it does:
https://doc.rust-lang.org/nightly/nightly-rustc/rustc_target/spec/abi/enum.Abi.html

However, this enum seems a lot more like it's describing "calling convention" than "ABI".